File: app/src/main/java/com/elewise/ldmobile/ui/FilterActivity.java
package com.elewise.ldmobile.ui;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.elewise.ldmobile.R;
import com.elewise.ldmobile.api.*;
import com.elewise.ldmobile.service.Session;
import com.elewise.ldmobile.utils.MessageUtils;
import com.elewise.ldmobile.widget.BaseWidget;
import com.elewise.ldmobile.widget.CheckboxWidget;
import com.elewise.ldmobile.widget.DateWidget;
import com.elewise.ldmobile.widget.InputWidget;
import com.elewise.ldmobile.widget.SelectWidget;

import java.util.ArrayList;
import java.util.List;

public class FilterActivity extends AppCompatActivity {
    private Button btnApply;
    private Button btnClear;
    private LinearLayout llDynamicPart;
    private List<BaseWidget> dynamicViewList = new ArrayList();

    private ProgressDialog progressDialog;
    AlertDialog dialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_filter);

        llDynamicPart = findViewById(R.id.llDynamicPart);
        btnApply = findViewById(R.id.btnApply);
        btnClear = findViewById(R.id.btnClear);

        btnClear.setOnClickListener(view -> {
            Session.getInstance().setFilterData(new FilterData[0]);
            finish();
        });

        btnApply.setOnClickListener(view -> {
            if (validateFilterData()) {
                Session.getInstance().setFilterData(getFilterData());
                finish();
            }
        });

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        progressDialog = new ProgressDialog(this);
        progressDialog.setCancelable(false);
        progressDialog.setMessage(getString(R.string.progress_dialog_load));
    }

    @Override
    protected void onResume() {
        super.onResume();
        loadData();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

    }

    private void loadData() {
        progressDialog.show();
        new Thread(() -> {
                    ParamFilterSettingsResponse
            response = Session.getInstance().getFilterSettings();
            handleFilterSettingsResponse(response);
        }).start();
    }

    private void handleFilterSettingsResponse(final ParamFilterSettingsResponse response) {
        runOnUiThread(() -> {
            progressDialog.cancel();

            String errorMessage = getString(R.string.error_load_data);

            if (response != null) {
                if (response.getStatus().equals(ResponseStatusType.S.name())) {
                    errorMessage = "";

                    for (FilterElement item : response.getFilters()) {
                        if (item.getType().equals("date")) {
                            DateWidget view = new DateWidget(this, item);
                            dynamicViewList.add(view);
                            llDynamicPart.addView(view);
                        } else if (item.getType().equals("string")) {
                            InputWidget view = new InputWidget(this, item);
                            dynamicViewList.add(view);
                            llDynamicPart.addView(view);
                        } else if (item.getType().equals("list")) {
                            SelectWidget view = new SelectWidget(this, item);
                            dynamicViewList.add(view);
                            llDynamicPart.addView(view);
                        } else if (item.getType().equals("checkbox")) {
                            CheckboxWidget view = new CheckboxWidget(this, item);
                            dynamicViewList.add(view);
                            llDynamicPart.addView(view);
                        } else {
                            Log.e("loadData", "uncnown filter type");
                        }
                    }
                } else if (response.getStatus().equals(ResponseStatusType.E.name())) {
                    if (!TextUtils.isEmpty(response.getMessage())) {
                        errorMessage = response.getMessage();
                    }
                } else if (response.getStatus().equals(ResponseStatusType.A.name())) {
                    errorMessage = getString(R.string.error_authentication);
                } else {
                    errorMessage = getString(R.string.error_unknown);
                    Log.e("getFilterSettings", "Unknown Status Type");
                }
            }

            if (!TextUtils.isEmpty(errorMessage)) {
                Toast.makeText(this, errorMessage, Toast.LENGTH_LONG).show();
            }
        });
    }

    @Override
    protected void onDestroy() {
        if (progressDialog != null) progressDialog.dismiss();
        if (dialog != null) dialog.dismiss();

        super.onDestroy();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == android.R.id.home) {
            finish();
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    private FilterData[] getFilterData() {
        ArrayList<FilterData> arrayList = new ArrayList();
        for (BaseWidget item : dynamicViewList) {
            if (!TextUtils.isEmpty(item.getValue1())) {
                arrayList.add(new FilterData(item.getName(), item.getValue1(), item.getValue2()));
            }
        }

        return arrayList.toArray(new FilterData[arrayList.size()]);
    }

    private Boolean validateFilterData() {
        for (BaseWidget item : dynamicViewList) {
            String value = item.validate();
            if (!TextUtils.isEmpty(value)) {
                MessageUtils.createDialog(this, getString(R.string.alert_dialog_error), getString(R.string.activity_filter_validate_error, value)).show();
                return false;
            }
        }
        return true;
    }

}

File: app/src/main/java/com/elewise/ldmobile/ui/FilterActivity.kt
package com.elewise.ldmobile.ui

import android.app.AlertDialog
import android.app.ProgressDialog
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import com.elewise.ldmobile.R
import com.elewise.ldmobile.api.FilterData
import com.elewise.ldmobile.api.ParamExecOperationActionType
import com.elewise.ldmobile.api.ParamFilterSettingsResponse
import com.elewise.ldmobile.api.ResponseStatusType
import com.elewise.ldmobile.service.Session
import com.elewise.ldmobile.utils.MessageUtils
import com.elewise.ldmobile.widget.*
import kotlinx.android.synthetic.main.activity_filter.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.util.*


class FilterActivity : BaseActivity() {
    private var viewModelJob = Job()
    private val uiScope = CoroutineScope(Dispatchers.Main + viewModelJob)
    var dialog: AlertDialog? = null

    private val dynamicViewList: MutableList<BaseWidget> = ArrayList()
    private var session = Session.getInstance()
    private var progressDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_filter)
        btnClear.setOnClickListener{ view: View? ->
            uiScope.launch {
                showProgressDialog()
                try {
                    val response = session.execOperation(ParamExecOperationActionType.reset_filters.name).await().body()
                    response?.let {
                        if (it.status == (ResponseStatusType.S.name)) {
                            loadData(true)
                        } else {
                            showError(getString(R.string.error_unknown))
                        }
                    }
                } catch (e: Exception) {
                    progressDialog?.dismiss()
                    showError()
                }
            }
        }
        btnApply.setOnClickListener{ view: View? ->
            if (validateFilterData()) {
                // fixme фильтр локально устанавливается тут, но на сервере только при отправке getDocuments
                session.filterData = filterData
                finish()
            }
        }
        llDynamicPart.setOnClickListener { hideKeyboard() }
        scrollView.setOnTouchListener { v, event ->
            hideKeyboard()
            false
        }
        updateActionBar(getString(R.string.filter_activity_title))
        loadData()
    }

    private fun loadData(withClean: Boolean = false) {
        uiScope.launch {
            try {
                showProgressDialog()
                val request = session.filterSettings
                val response = request.await()
                handleFilterSettingsResponse(response.body(), withClean)
            } catch (e: Exception) {
                progressDialog?.dismiss()
                showError()
            }
        }
    }

    private fun showError(errorMessage: String = getString(R.string.error_load_data)) {
        // показать ошибку
        dialog = MessageUtils.createDialog(this, getString(R.string.alert_dialog_error), errorMessage).apply { show() }
    }

    private fun showProgressDialog() {
        if (progressDialog == null) {
            progressDialog = ProgressDialog(this).apply {
                setCancelable(false)
                setMessage(getString(R.string.progress_dialog_load))
            }
        }
        progressDialog?.show()
    }

    private fun handleFilterSettingsResponse(response: ParamFilterSettingsResponse?, withClean: Boolean = false) {
        progressDialog?.cancel()
        if (response != null) {
            if (response.status == ResponseStatusType.S.name) {
                llDynamicPart.removeAllViews()
                dynamicViewList.clear()
                // в цикле добавляем виджеты
                response.filters?.let {
                    for (item in it) {
                        if (item.type == "date") {
                            val view = DateWidget(this, item)
                            dynamicViewList.add(view)
                            llDynamicPart.addView(view)
                        } else if (item.type == "string") {
                            val view = InputWidget(this, item)
                            dynamicViewList.add(view)
                            llDynamicPart.addView(view)
                        } else if (item.type == "list") {
                            val view = SelectWidget(this, item)
                            dynamicViewList.add(view)
                            llDynamicPart.addView(view)
                        } else if (item.type == "checkbox") {
                            val view = CheckboxWidget(this, item)
                            dynamicViewList.add(view)
                            llDynamicPart.addView(view)
                        } else {
                            Log.e("loadData", "unknown filter type")
                        }
                    }
                }
                if (withClean) {
                    // сохраним дефольтный фильтр
                    session.filterData = filterData
                }
            } else if (response.status == ResponseStatusType.E.name) {
                if (response.message?.isNotEmpty() == true) {
                    showError(response.message)
                } else {
                    showError()
                }
            } else if (response.status == ResponseStatusType.A.name) {
                session.errorAuth()
            } else {
                showError(getString(R.string.error_unknown_status_type))
            }
        } else {
            showError()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModelJob.cancel()
        progressDialog?.dismiss()
        dialog?.dismiss()
    }

    private val filterData: List<FilterData>
        get() {
            val arrayList: ArrayList<FilterData> = ArrayList()
            for (item in dynamicViewList) {
                arrayList.add(FilterData(item.getName(), item.getValue1(), item.getValue2()))
            }
            return arrayList
        }

    private fun validateFilterData(): Boolean {
        for (item in dynamicViewList) {
            val value = item.validate()
            if (!TextUtils.isEmpty(value)) {
                MessageUtils.createDialog(this, getString(R.string.alert_dialog_error), getString(R.string.activity_filter_validate_error, value)).show()
                return false
            }
        }
        return true
    }

    fun hideKeyboard() {
        dynamicViewList.forEach {
            if (it is InputWidget) {
                it.clearFocusEditView()
            }
        }
        hideKeyboard(llDynamicPart)
    }
}